File: _40_combination_sum_ii/main.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []
        candidates.sort()

        def backtracking(i, current, subset):
            nonlocal result

            if current == 0:
                result.append(subset.copy())
                return

            if i >= len(candidates) or current < 0:
                return

            subset.append(candidates[i])
            current -= candidates[i]
            backtracking(i + 1, current, subset)

            while i + 1 < len(candidates) and candidates[i] == candidates[i + 1]:
                i += 1

            subset.pop()
            current += candidates[i]
            backtracking(i + 1, current, subset)


        backtracking(0, target, [])

        return result

File: _40_combination_sum_ii/main.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<int> current;
        backtracking(candidates, target, 0, current);
        return m_Result;
    }

    void backtracking(vector<int>& candidates, int target, int i, vector<int>& current)
    {
        if (target == 0)
        {
            m_Result.push_back(current);
            return;
        }

        if (i == candidates.size() || target < 0)
        {
            return;
        }

        current.push_back(candidates[i]);
        target -= candidates[i];
        backtracking(candidates, target, i + 1, current);

        // While is the same number
        while (i + 1 < candidates.size() && candidates[i] == candidates[i + 1])
        {
            i += 1;
        }

        current.pop_back();
        target += candidates[i];
        backtracking(candidates, target, i + 1, current);
    }

private:
    vector<vector<int>> m_Result;
};